AP_Mission: Added MIS_RETAIN Parameter to clear Mission on boot Issue…

ChrisBird opened the pull request:
… #5743 & #6383 (closed)

Added the MIS_RETAIN parameter to allow people to clear a mission on boot up if they wish.  It's default is to use current behavior which is to retain any missions that are loaded.  This has been tested on SITL (both positive and negative scenarios) but not a read device.  I can test it on either a Pixhawk or Cube (Pixhawk 2.1) if required.

As this is my first change to Ardupilot, please ensure I've followed the sytle guide correctly.

Let me know if I need to do anything extra....

khancyr commented:
Hello, 
The code style is good !
The changes look good to me too

tridge commented on libraries/AP_Mission/AP_Mission.cpp:
```diff
@@ -22,6 +22,13 @@ const AP_Param::GroupInfo AP_Mission::var_info[] = {
+    // @Description: Controls mission if the mission is retained after a reboot, 0 to clear, 1 to retain.
+    // @Values: 0:Clear Mission, 1:Retain Mission
+    // @User: Advanced
+    AP_GROUPINFO("RETAIN",  2, AP_Mission, _retain, AP_MISSION_RETAIN_DEFAULT),
```
maybe MIS_CLEAR_BOOT as a clearer parameter name?

tridge commented:
on dev call, we thought that making it MIS_OPTIONS would be a good choice, allowing it to be a bitmask of options in the future. If you need help implementing that change please let me know - I'd be happy to do it based on your PR

ChrisBird commented:
Hi @tridge, I'm more than happy to implement the changes above as I'm using this to get more familiar with the source code and help contribute something that I've gotten a lot of use out of.  I'll give it a go today will submit changes later today.

magicrub commented:
Great, thanks @ChrisBird . Welcome to the project!

ChrisBird commented:
Hi @tridge, I've updated it to use the parameter MIS_OPTIONS as a bitmask as requested.

Hi @magicrub, thank you!

WickedShell commented on libraries/AP_Mission/AP_Mission.h:
```diff
@@ -504,6 +512,7 @@ class AP_Mission {
     // parameters
     AP_Int16                _cmd_total;  // total number of commands in the mission
     AP_Int8                 _restart;   // controls mission starting point when entering Auto mode (either restart from beginning of mission or resume from last command run)
+    AP_Int8                _options;    // bitmask options for missions, currently for mission retention on reboot but can be expanded as required
```
If this is going to be an options bit mask, I'd rather we store it in a AP_Int16 at least so that we have a bit more in the way of expansion available. (AP_Int32 is a bit overrated, as the float transport mechanism sets an early limit)

OXINARF replied:
I agree, 16 is better.

WickedShell commented on libraries/AP_Mission/AP_Mission.cpp:
```diff
@@ -45,6 +52,12 @@ void AP_Mission::init()
     }
+    
+    // If Mission retain bit is not set then it should clear the mission
+    if(!is_option(MASK_MISSION_RETAIN)) {
```
`!is_option()` doesn't seem like it's actually cleaner then just doing `mask & MISSION_MASK_CLEAR` (making the bit set to flag clear on reboot would also simplify the code at runtime without any user comprehension problems).

OXINARF replied:
I agree, I don't see the need for the method and the MASK should be CLEAR.

OXINARF commented on libraries/AP_Mission/AP_Mission.cpp:
```diff
@@ -22,6 +22,13 @@ const AP_Param::GroupInfo AP_Mission::var_info[] = {
+    // @Param: OPTIONS
+    // @DisplayName: Mission options bitmask
+    // @Description: Bitmap of what options to use. This values is made up of the sum of each of the mission options you require.  The individual bits are CLEAR_MISSION=1
+    // @Values: 0:Clear Mission, 1:Retain Mission
```
This should be a @<span></span>Bitmask, not Values. Also, 0 should be current behavior (retain mission) and 1 clear mission.

OXINARF commented on libraries/AP_Mission/AP_Mission.cpp:
```diff
@@ -22,6 +22,13 @@ const AP_Param::GroupInfo AP_Mission::var_info[] = {
 
+    // @Param: OPTIONS
+    // @DisplayName: Mission options bitmask
+    // @Description: Bitmap of what options to use. This values is made up of the sum of each of the mission options you require.  The individual bits are CLEAR_MISSION=1
```
This values is -> The value is

ChrisBird commented:
I've removed the is_clear method, moved the logic to the init method. I've updated it to a 16 bit bitmask, updated the comments and switched the @values to @bitmask.  Tested and uploaded.

OXINARF commented:
I've squashed everything up and changed a couple of things:

- commit message
- parameter description (also added *on reboot* to the bitmask value)
- added *AP_MISSION_* to the MASK macro

@ChrisBird Please check it is OK to you: https://github.com/OXINARF/ardupilot/commit/32914ed2fff03283cc6f8c65d362d10a7f542483
@WickedShell Can you give a look too? Thanks.

ChrisBird commented:
@OXINARF had a look and I'm ok with it.

WickedShell commented:
@OXINARF LGTM, I think we will always have run that before the mission can be considered to be running, so ignoring the return value from clear() should be fine.

OXINARF commented:
@WickedShell I agree that it shouldn't be a problem with the current `clear` code. I actually thought about what we would do if it didn't clear the mission and I see no good option. Continue trying, panicking? Both seem too extreme to me.

@ChrisBird Merged, thanks!

WickedShell commented:
@OXINARF Yeah, I only pointed it out this morning because until a fresh cup of coffee I hadn't thought of it. The problem with panic is that I think it creates a situation where you can't change the parameter to fix the behavior, otherwise I'd agree panic is the logical approach.
